dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -531,6 +531,272 @@ template class NaryOpNode<functional::min<double>>;
+
+    void revert() {
+        ArrayNodeStateData::revert();
+        extra.revert();
+    }
+
+    // op-specific storage.
+    ExtraArrayData<BinaryOp> extra;
+};
+
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* node_ptr, ssize_t axis, double init)
+        : ArrayOutputMixin(partial_reduce_shape(node_ptr->shape(), axis)),
+          init(init),
+          array_ptr_(node_ptr),
+          axis_(axis),
+          parent_strides_c_(as_contiguous_strides(array_ptr_->shape())) {
+    if (array_ptr_->dynamic()) {
+        throw std::invalid_argument("cannot do a partial reduction on a dynamic array");
+    } else if (array_ptr_->size() < 1) {
+        throw std::invalid_argument("cannot do a partial reduction on an empty array");
+    }
+
+    /// TODO: support negative axis
+    if (axis_ < 0 || axis_ >= array_ptr_->ndim()) {
+        throw std::invalid_argument("Axis should be an integer between 0 and n_dim - 1");
+    }
+
+    add_predecessor(node_ptr);
+}
+
+template <>
+PartialReduceNode<std::plus<double>>::PartialReduceNode(ArrayNode* array_ptr, ssize_t axis)
+        : PartialReduceNode(array_ptr, axis, 0) {}
+
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* array_ptr, ssize_t axis)
+        : init(), array_ptr_(array_ptr), axis_(axis) {
+    if (array_ptr_->dynamic()) {
+        throw std::invalid_argument(
+                "cannot do a reduction on a dynamic array with an operation that has no identity "
+                "without supplying an initial value");
+    } else if (array_ptr_->size() < 1) {
+        throw std::invalid_argument(
+                "cannot do a reduction on an empty array with an operation that has no identity "
+                "without supplying an initial value");
+    }
+
+    /// TODO: support negative axis
+    if (axis_ < 0 || axis_ >= array_ptr_->ndim()) {
+        throw std::invalid_argument("Axis should be an integer between 0 and n_dim - 1");
+    }
+
+    add_predecessor(array_ptr);
+}
+
+template <class BinaryOp>
+void PartialReduceNode<BinaryOp>::commit(State& state) const {
+    data_ptr<PartialReduceNodeData<op>>(state)->commit();
+}
+
+template <class BinaryOp>
+double const* PartialReduceNode<BinaryOp>::buff(const State& state) const {
+    return data_ptr<PartialReduceNodeData<op>>(state)->buff();
+}
+
+template <class BinaryOp>
+std::span<const Update> PartialReduceNode<BinaryOp>::diff(const State& state) const {
+    return data_ptr<PartialReduceNodeData<op>>(state)->diff();
+}
+
+template <class BinaryOp>
+void PartialReduceNode<BinaryOp>::initialize_state(State& state) const {
+    int index = topological_index();
+    assert(index >= 0 && "must be topologically sorted");
+    assert(static_cast<int>(state.size()) > index && "unexpected state length");
+    assert(state[index] == nullptr && "already initialized state");
+
+    std::vector<double> values(size(state));
+
+    for (ssize_t i = 0; i < size(state); ++i) {
```

Comment:
```suggestion
    for (ssize_t i = 0, stop = size(state); i < stop; ++i) {
```
See https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop. Not quite the same, but the principal applies. Especially since `size()` is a virtual function.